src/topology/mod.rs

```diff
@@ -29,11 +32,26 @@ pub struct RunningTopology {
     abort_tx: mpsc::UnboundedSender<()>,
 }
 
+#[derive(PartialEq)]
+pub enum Stage {
+    ConfigValidated,
+    HealthChecksPassed,
+}
+
 pub fn start(
-    config: Result<Config, Vec<String>>,
+    config: Config,
     rt: &mut tokio::runtime::Runtime,
     require_healthy: bool,
 ) -> Option<(RunningTopology, mpsc::UnboundedReceiver<()>)> {
+    start_or_validate(config, rt, None, require_healthy).right()
+}
+
+pub fn start_or_validate(
+    config: Config,
+    rt: &mut tokio::runtime::Runtime,
+    exit_after: Option<Stage>,
+    require_healthy: bool,
+) -> Either<bool, (RunningTopology, mpsc::UnboundedReceiver<()>)> {
```

Comment:
Can this just be a `Option<(...)>` where the `None` case represents a failure. This could also be a `Result` I feel that `Either` doesn't make much sense here.

Reply:
Well, that's the whole point.. Left side here is used for the case, when we do not have `RunningTopology` and after inroducing `--dry-run` this doesn't always mean error. So, `Result` can be confusing here, because `Error` would be used for legal exit with no errors. `Option` is not enought, because we use boolean flag to exit with either 0 or 78 exit code.

Reply:
I think this is a good sign that this function has gotten a little too confusing :smile: 

Could we break it up into separate `validate` and `start` functions, where `start` would first call `validate`?

Reply:
This is a good idea. As far I remember, I wanted to do something like this but got some issues.
I'll revise it tomorrow. `Either<bool` looks a bit confusing for me too, though it is a least evil in certain circumstances.

Reply:
Hey guys, check out new version, I think it became simpler.